prov/usnic/src/usdf_poll.c

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright (c) 2016, Cisco Systems, Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fi.h>
+#include <fi_enosys.h>
+#include <fi_util.h>
+
+#include "usdf.h"
+#include "usdf_cq.h"
+#include "usdf_poll.h"
+
+static int usdf_poll_poll(struct fid_poll *fps, void **context, int count)
+{
+	struct usdf_cq *cq;
+	struct usdf_poll *ps;
+	struct dlist_entry *item;
+	struct fid_list_entry *entry;
+	int ret = 0;
+	int copied = 0;
+	int pending;
+
+	USDF_TRACE_SYS(DOMAIN, "\n");
+
+	if (!fps || !context) {
+		USDF_WARN_SYS(DOMAIN, "pollset and context can't be NULL.\n");
+		return -FI_EINVAL;
+	}
+
+	ps = poll_ftou(fps);
+
+	usdf_domain_progress(ps->poll_domain);
+
+	fastlock_acquire(&ps->lock);
+
+	dlist_foreach(&ps->list, item) {
+		entry = container_of(item, struct fid_list_entry, entry);
+		assert(entry->fid->fclass == FI_CLASS_CQ);
+
+		cq = cq_fidtou(entry->fid);
+
+		pending = 0;
+		if (usdf_cq_is_soft(cq)) {
+			if (cq->c.soft.cq_tail == cq->c.soft.cq_head) {
+				/* Make sure it isn't. */
+				pending = (cq->c.soft.cq_last_op !=
+					USDF_SOFT_CQ_READ);
+			} else {
+				pending = 1;
+			}
+		} else {
+			if (!usdf_check_empty_hard_cq(cq->c.hard.cq_cq))
```

Comment:
> Why no corresponding usdf_check_empty_soft_cq?

I could do that but it's really easy to access the parts of the CQ to tell whether it is empty or not. The reason I made one for the hard cq is because it involves a lot of unraveling. I'll create the appropriate one for soft CQ as well.

> Also, not sure what Make sure it isn't means here...

Oops, forgot to finish that comment! It was supposed to say "Make sure it isn't empty".